chronos/asyncmacro2.nim

```diff
@@ -242,19 +235,26 @@ proc asyncSingleProc(prc: NimNode): NimNode {.compileTime.} =
         newNimNode(nnkBracketExpr, prc)
         .add(newIdentNode("Future"))
         .add(newIdentNode("void"))
+        .add(newIdentNode("void"))
   if procBody.kind != nnkEmpty:
     result.body = outerProcBody
   #echo(treeRepr(result))
   #if prcName == "recvLineInto":
   #  echo(toStrLit(result))
 
-template await*[T](f: Future[T]): untyped =
+macro checkMagical(f, e: typed): untyped =
+  result = newNimNode(nnkStmtList)
+
+  let types = getType(e)
+  if types.len < 1: return
+  for errorType in types[1..^1]:
+    result.add quote do:
+      if not isNil(`f`.error) and `f`.error is `errorType`:
```

Comment:
The `is` operator works at compile-time. It will compare the [type-erased](https://en.wikipedia.org/wiki/Type_erasure) `error` field to the specified exception types.  To correct the code, you need to use the `of` operator which will work at run-time by examining the concrete exception type that ended being stored in the `error` field.

Having that said, it's a bit unnecessary to go through the trouble of doing these run-time checks. You only need to fool the compiler that the expression can raise all these exceptions even when you directly `raise f.error` without any casting. 

Theoretically, you should be able to do that with a `cast(raises: [...])` block:

```nim
{.cast(raises: [ValueError]).}:
  echo foo
  echo bar
```

... but this requires Nim 1.6. You can also fool the compiler by putting the raises statements in an arbitrary block of code that appears "reachable" to the compiler (even when it's not actually reachable). Accidentally, this is exactly what you did here by using the `is` operator.

Reply:
Oh, right, my bad, should have used of

Though in 1.2, I don't see a better way, not only I have to tell the compiler that this expression can raise E errors, but also that it can't raise anything else

I couldn't find a way to force the raises effect in 1.2, like you did